libraries/node-core-library/src/LockFile.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fsx from 'fs-extra';
+import * as path from 'path';
+import * as child_process from 'child_process';
+
+export function getProcessStartTime(pid: string): string | undefined {
+  let args: string[];
+  if (process.platform === 'darwin') {
+    args = [`-p ${pid}`, '-f'];
+  } else if (process.platform === 'linux') {
+    args = ['-p', pid.toString(), '-f'];
+  } else {
+    throw new Error(`Unsupported system: ${process.platform}`);
+  }
+
+  const psResult: string = child_process.spawnSync('ps', args).stdout.toString();
+  const psSplit: string[] = psResult.split('\n');
+
+  if (psSplit[1]) {
+    // note that on OSX the first entry is a ""
+    // UID, PID, PPID, C, STIME, TTY, TIME, CMD
+    // e.g.: items = ['root', '2848', '1', '0', '13:08', '?', '00:00:00', '/usr/lib/udisks2/udisksd']
+    const items: string[] = psSplit[1].split(/\s+/);
+    return items[process.platform === 'darwin' ? 5 : 4];
+  }
+  return undefined;
+}
+
+/**
+ * A helper utility for working with file-based locks.
+ * This class should only be used for locking resources across processes,
+ * but should not be used for attempting to lock a resource in the same process.
+ * @public
+ */
+export class LockFile {
+  /**
+   * Attempts to create a lockfile with the given filePath.
+   * If successful, returns a LockFile instance.
+   * If unable to get a lock, returns undefined.
+   * @param filePath - the filePath of the lockfile.
+   */
+  public static tryAcquire(filePath: string): LockFile | undefined {
+    let dirtyWhenAcquired: boolean = false;
+
+    if (fsx.existsSync(filePath)) {
+
+      dirtyWhenAcquired = true;
+
+      if (process.platform === 'win32') {
+        // If the lockfile is held by an process with an exclusive lock, then removing it will
+        // silently fail. OpenSync() below will then fail and we will be unable to create a lock.
+
+        // Otherwise, the lockfile is sitting on disk, but nothing is holding it, implying that
+        // the last process to hold it died.
+      } else if (process.platform === 'darwin' || process.platform === 'linux') {
+        // we need to compare the process ID and start time in the lockfile
+        // to the result of the ps command
+
+        const contents: string = fsx.readFileSync(filePath).toString();
+        const [pid, startTime] = contents.split(';');
+
+        if (!!pid && !!startTime) {
```

Comment:
There should be unit test cases for empty file, successfully parsed, and failed to parse

Reply:
damnit

---
In reply to: [165796326](https://github.com/Microsoft/web-build-tools/pull/517#discussion_r165796326) [](ancestors = 165796326)